### PR TITLE
Add a 'max_send_frag' option to configure maximum size of send fragments

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
#### Checklist
- [X] documentation is added or updated
- [ ] tests are added or updated

Noticed while working on #1008

### CODE REVIEW

File: apps/s_client.c [hunk 1]
```diff
@@ -1432,6 +1442,11 @@ int s_client_main(int argc, char **argv)
         goto end;
     }
 
+    if (max_send_fragment > SSL3_RT_MAX_PLAIN_LENGTH) {
+        BIO_printf(bio_err, "Bad max send fragment size\n");
```
Comment (richsalz) [inline apps/s_client.c]: plug put "%s: " and print prog.

File: apps/s_server.c [hunk 2]
```diff
@@ -1543,6 +1556,11 @@ int s_server_main(int argc, char *argv[])
     }
 #endif
 
+    if (max_send_fragment > SSL3_RT_MAX_PLAIN_LENGTH) {
+        BIO_printf(bio_err, "Bad max send fragment size\n");
```
Comment (richsalz) [inline apps/s_server.c]: output prog name here, too.  I know it's not always done, but let's get in the habit of fixing these.

File: doc/man1/s_client.pod [hunk 3]
```diff
@@ -398,6 +399,11 @@ asynchronously. This will only have an effect if an asynchronous capable engine
 is also used via the B<-engine> option. For test purposes the dummy async engine
 (dasync) can be used (if available).
 
+=item B<-max_send_frag int>
+
+The maximum size of data fragment to send.  See
+L<SSL_CTX_set_max_send_fragment(3)> for further information.
```
Comment (richsalz) [inline doc/man1/s_client.pod]: If there's a SEE ALSO section, add a link there, too.

File: apps/s_server.c (hunk 2 above)
Comment (FdaSilvaYY) [inline apps/s_server.c]: I will fiw some more in another PR ;)

File: doc/man1/s_client.pod [hunk 4]
```diff
@@ -398,6 +399,11 @@ asynchronously. This will only have an effect if an asynchronous capable engine
 is also used via the B<-engine> option. For test purposes the dummy async engine
 (dasync) can be used (if available).
 
+=item B<-max_send_frag int>
+
+The maximum size of data fragment to send.
+See L<SSL_CTX_set_max_send_fragment(3)> for further information.
```
Comment (dot-asm) [inline doc/man1/s_client.pod]: Well, SSL_CTX_set_max_send_fragment says that it "will only accept a value in the range 512 - SSL3_RT_MAX_PLAIN_LENGTH." This kind of suggests that one should check for it's return value and at least issue warning that attempt is failed.

Comment (dot-asm) [inline doc/man1/s_client.pod]: I mean as it stands now it only checks for upper limit.

File: apps/s_server.c [hunk 5]
```diff
@@ -1741,6 +1759,10 @@ int s_server_main(int argc, char *argv[])
     if (async) {
         SSL_CTX_set_mode(ctx, SSL_MODE_ASYNC);
     }
+
+    if (max_send_fragment > 0)
+        SSL_CTX_set_max_send_fragment(ctx, max_send_fragment);
```
Comment (levitte) [inline apps/s_server.c]: The returned value should be checked.  0 indicates failure...  and really, that should also be done for the three calls below as well...

File: doc/man1/s_client.pod (hunk 4 above)
Comment (FdaSilvaYY) [inline doc/man1/s_client.pod]: Checks added !

File: apps/s_client.c [hunk 6]
```diff
@@ -1432,13 +1428,20 @@ int s_client_main(int argc, char **argv)
...
+         || split_send_fragment > SSL3_RT_MAX_PLAIN_LENGTH ) {
+        BIO_printf(bio_err, "%s: Invalid split send fragment size\n", prog);
         goto end;
     }
 
     if (max_pipelines > SSL_MAX_PIPELINES) {
-        BIO_printf(bio_err, "Bad max pipelines value\n");
+        BIO_printf(bio_err, "%s: too large max-pipelines value\n", prog);
         goto end;
     }
```
Comment (levitte) [inline apps/s_client.c]: Ok, all these checks are a bit over the top.  What happens when the libssl code changes those ranges, will we have to keep track of that and update the values here as well?  It should be enough that `SSL_CTX_set_max_send_fragment()`, `SSL_CTX_set_split_send_fragment()`, `SSL_CTX_set_max_pipelines()` and SSL_ friends return 0 when the given argument is out of range...

File: apps/s_client.c [hunk 7]
```diff
@@ -1563,11 +1566,23 @@ int s_client_main(int argc, char **argv)
     if (async) {
         SSL_CTX_set_mode(ctx, SSL_MODE_ASYNC);
     }
-    if (split_send_fragment > 0) {
-        SSL_CTX_set_split_send_fragment(ctx, split_send_fragment);
+
+    if (max_send_fragment > 0
+        && !SSL_CTX_set_max_send_fragment(ctx, max_send_fragment)) {
+        BIO_printf(bio_err, "%s: Error setting max send fragment size\n", prog);
```
Comment (levitte) [inline apps/s_client.c]: I suggest this:

``` C
    if (!SSL_CTX_set_max_send_fragment(ctx, max_send_fragment)) {
        BIO_printf(bio_err, "%s: Max send fragment size %u is out of permitted range\n",
                   prog, max_send_fragment);
```

File: apps/s_client.c [hunk 8]
```diff
@@ -1563,11 +1566,23 @@ int s_client_main(int argc, char **argv)
...
+        goto end;
     }
-    if (max_pipelines > 0) {
-        SSL_CTX_set_max_pipelines(ctx, max_pipelines);
+
+    if (split_send_fragment > 0
+        && !SSL_CTX_set_split_send_fragment(ctx, split_send_fragment)) {
+        BIO_printf(bio_err, "%s: Error setting split send fragment size\n", prog);
```
Comment (levitte) [inline apps/s_client.c]: I suggest this:

``` C
    if (!SSL_CTX_set_split_send_fragment(ctx, split_send_fragment)) {
        BIO_printf(bio_err, "%s: Split send fragment size %u is out of permitted range\n",
                   prog, split_send_fragment);
```

File: apps/s_client.c [hunk 9]
```diff
@@ -1563,11 +1566,23 @@ int s_client_main(int argc, char **argv)
...
+        goto end;
+    }
+
+    if (max_pipelines > 0
+        && !SSL_CTX_set_max_pipelines(ctx, max_pipelines)) {
+        BIO_printf(bio_err, "%s: Error setting max pipelines\n", prog);
```
Comment (levitte) [inline apps/s_client.c]: I suggest this:

``` C
    if (!SSL_CTX_set_max_pipelines(ctx, max_pipelines)) {
        BIO_printf(bio_err, "%s: Max pipelines %u is out of permitted range\n",
                   prog, max_pipelines);
```

File: apps/s_server.c [hunk 10]
```diff
@@ -1543,13 +1542,19 @@ int s_server_main(int argc, char *argv[])
...
+    if ((split_send_fragment > 0 && split_send_fragment < 512)
+         || split_send_fragment > SSL3_RT_MAX_PLAIN_LENGTH ) {
+        BIO_printf(bio_err, "%s: Invalid split send fragment size\n", prog);
+        goto end;
+    }
     if (max_pipelines > SSL_MAX_PIPELINES) {
-        BIO_printf(bio_err, "Bad max pipelines value\n");
+        BIO_printf(bio_err, "%s:too large max-pipelines value\n", prog);
         goto end;
     }
```
Comment (levitte) [inline apps/s_server.c]: Same comment here as for `s_client.c`

File: apps/s_server.c [hunk 11]
```diff
@@ -1741,11 +1746,22 @@ int s_server_main(int argc, char *argv[])
...
+        && !SSL_CTX_set_split_send_fragment(ctx, split_send_fragment)) {
+        BIO_printf(bio_err, "%s: Error setting split send fragment size\n", prog);
+        goto end;
     }
-    if (max_pipelines > 0) {
-        SSL_CTX_set_max_pipelines(ctx, max_pipelines);
+    if (max_pipelines > 0
+        && !SSL_CTX_set_max_pipelines(ctx, max_pipelines)) {
+        BIO_printf(bio_err, "%s: Error setting max pipelines value\n", prog);
+        goto end;
```
Comment (levitte) [inline apps/s_server.c]: Same suggestions here as in `s_client.c`

**Comment by FdaSilvaYY** [comment]
@levitte : rebased and error message updated.

**Comment by levitte** [comment]
I think you misunderstood me.  It wasn't just about the message, but also about all that unnecessary checking of values, which is done in `SSL_CTX_set_max_send_fragment()` et al anyway!

**Comment by FdaSilvaYY** [comment]
@levitte : rebased on top of last master. It it OK ?

File: apps/s_client.c [hunk 12]
```diff
@@ -1444,6 +1440,7 @@ int s_client_main(int argc, char **argv)
         goto end;
     }
 
+<<<<<<< HEAD
```
Comment (levitte) [inline apps/s_client.c]: I think this line should go ;-)

Comment (FdaSilvaYY) [inline apps/s_client.c]: Gone ! ;)

**Comment by FdaSilvaYY** [comment]
Rebased and conflict fixed. Ready to merge .

**Comment by levitte** [comment]
Merged.  28e5ea88a8fe93aa6dd023a5abd54f01c1ad6450 to 36b2cfb191f4a3f3a49b7c4d865308b7021daa42

**Comment by richsalz** [comment]
should the doc say that the RELEASE-BUFFERS flag is needed for this to actually reduce memory?

**Pull request closed without merging**